File: StriverA2Z/Java/BinaryTrees/MaximumWidth.java
import java.util.LinkedList;
import java.util.Queue;

public class MaximumWidth {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[]{1, 2, 3, 4, 5, 6});
        System.out.println(maximumWidth(root));
    }

    public static int maximumWidth(Node root) {
        int ans = 0;
        if (root == null) return ans;
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(root, 0));
        int last = 0, first = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            int min = q.peek().num;
            for (int i = 0; i < size; i++) {
                int curr = q.peek().num - min;
                Node node = q.peek().node;
                q.poll();
                if (i == 0) first = curr;
                if (i == size - 1) last = curr;
                if (node.left != null) q.offer(new Pair(node.left, 2 * curr + 1));
                if (node.right != null) q.offer(new Pair(node.right, 2 * curr + 2));
            }
            ans = Math.max(ans, last - first + 1);
        }
        return ans;
    }
}

File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/maximumWidth.ts
import {arrayToBTLevelOrder, BTNode} from "./BinaryTreeOperations";

function maximumWidth(root: BTNode) {
	let ans = 0;
	if (!root) return ans;
	let q: {node: BTNode, num: number}[] = [];
	q.push({node: root, num: 0});
	let first, last;
	while (q.length > 0) {
		let size = q.length;
		let min = q[0].num;
		for (let  i = 0; i < size; i++) {
			let {node, num} = q[0];
			num -= min;
			q.shift();
			if (i === 0) first = num;
			if (i === size - 1) last = num;
			if (node.left) q.push({node: node.left, num: 2 * num + 1});
			if (node.right) q.push({node: node.right, num: 2 * num + 2});
		}
		ans = Math.max(ans, last - first + 1);
	}
	return ans;
}

let root = arrayToBTLevelOrder([1,2,3,4,5,6]);
console.log(maximumWidth(root));